### ISSUE TITLE
uploading file via boto3 resource api locks up minio for 60 seconds

### ISSUE THREAD

**Issue by abulka** [minio/minio#11950]
## Expected Behavior

Using the python boto3 api, uploading a tiny file to minio should be very fast, and not freeze minio (or boto3?) for 60 seconds. 

This same code, hitting Amazon, works ok - minio should work the same as Amazon s3.

## Current Behavior

When there are two calls using boto3 resource api, one for creating a simple file via `put`, the second for an `upload_file`, the upload call causes minio (or boto3) to freeze for 60 seconds, then the file is uploaded a second time, unecessarily. 

```python
s3resource.Object(BUCKET, 'storage/file1.css').put(Body='')
s3resource.Bucket(BUCKET).upload_file(PATH_TO_PNG_FILE_ON_DISK, FILENAME_ON_S3)
```

## Possible Solution

Using the boto3 client api `put_object` (instead of the resource api `put`) for the first file creation allows the upload to complete instantly.

```python
s3client.put_object(Bucket=BUCKET, Key='storage/file1.css') # using s3client instead
s3resource.Bucket(BUCKET).upload_file(PATH_TO_PNG_FILE_ON_DISK, FILENAME_ON_S3) # now works OK
```

Alternatively, uploading the file via the boto3 client api allows the upload to complete instantly, even if the previous file creation was done via the resource api

```python
s3resource.Object(BUCKET, 'storage/file1.css').put(Body='') # uses resource api
s3client.upload_file(PATH_TO_PNG_FILE_ON_DISK, BUCKET, FILENAME_ON_S3)  #  now works OK
```

## Steps to Reproduce (for bugs)

1. Create a bucket on minio
2. Run this code

```python
import boto3

BUCKET = 'some.bucket'
HTTP = 'http://'
ENDPOINT = 'localhost:9000'
AWS_ACCESS_KEY_ID = 'minioadmin'
AWS_SECRET_ACCESS_KEY = 'minioadmin'
REGIONNAME = 'whatever'

FILENAME_ON_S3 = 'storage/file1.png'
PATH_TO_PNG_FILE_ON_DISK = 'static/images/image1.png'
MSG = 'uploading png file...'

session = boto3.Session(
    aws_access_key_id=AWS_ACCESS_KEY_ID,
    aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
    aws_session_token=None,
    region_name=REGIONNAME,
    botocore_session=None,
    profile_name=None)

s3client = session.client(
    's3', endpoint_url=f"{HTTP}{ENDPOINT}")

s3resource = session.resource(
    's3', endpoint_url=f"{HTTP}{ENDPOINT}")

# this call must be done first in order for bug to happen
s3resource.Object(BUCKET, 'storage/file1.css').put(Body='')

print(MSG)
# locks up for 60 secs on minio?!
s3resource.Bucket(BUCKET).upload_file(PATH_TO_PNG_FILE_ON_DISK, FILENAME_ON_S3)

print('uploading complete')
```

This freeze does not happen when the endpoint is Amazon s3.

Minio logs via `mc admin trace -a minio`

```
23:35:47.541 [200 OK] s3.PutObject localhost:9000/mybucket/storage/file1.css                   1.576ms      ↑ 112 B ↓ 230 B
23:35:47.548 [200 OK] s3.PutObject localhost:9000/mybucket/storage/file1.png                   2.165ms      ↑ 77 KiB ↓ 230 B
23:36:47.665 [200 OK] s3.PutObject localhost:9000/mybucket/storage/file1.png                   4.808ms      ↑ 77 KiB ↓ 230 B
```

The first two lines happen immediately, then the 60 second freeze happens, followed by the third line.

Speculating, it looks like minio erroneously thinks there is a error, waits *exactly 60 seconds* (look at the time stamps) then uploads the `.png` file again? Or maybe its `boto3` doing the wait and re-upload, because boto3 isn't getting the correct response from minio? Remember, this same code runs ok targeting the Amazon s3 endpoint.

## Context

Am using minio s3 for developing my Python 3.8 application, switching to Amazon s3 endpoint for production. Expecting minio to work the same and mixing and matching s3 client and s3 resource api calls as needed.

## Your Environment

* minio version RELEASE.2021-03-17T02-33-02Z
* Server setup and configuration: python 3.8 running a simple script, no server needed to repro
* Operating System and version (`uname -a`): Mac Mojave, Darwin mymac.local 18.7.0 Darwin Kernel Version 18.7.0: Tue Jan 12 22:04:47 PST 2021; root:xnu-4903.278.56~1/RELEASE_X86_64 x86_64 i386 iMac17,1 Darwin

**Comment by harshavardhana** [comment]
This is a known boto3 issue and they have ignored our requests to fix it @abulka https://github.com/boto/boto/issues/840

**Comment by harshavardhana** [comment]
> Speculating, it looks like minio erroneously thinks there is a error, waits _exactly 60 seconds_ (look at the time stamps) then uploads the `.png` file again? Or maybe its `boto3` doing the wait and re-upload, because boto3 isn't getting the correct response from minio? Remember, this same code runs ok targeting the Amazon s3 endpoint.

Its not MinIO its boto3 that is not implementing 100-continue correctly.

**Comment by abulka** [comment]
@harshavardhana Interesting, thanks. This of course makes using `boto3` + `minio` s3 unreliable. I just ran into an even simpler repro case of the 60 sec freeze - creating an empty file followed by file with some content:

```python
s3resource.Object(BUCKET, 'storage/file1.css').put(Body='')
s3resource.Object(BUCKET, 'storage/file2.json').put(Body='{}')
```

- Note that creating two empty files with content e.g. body '' works OK
-  Note that creating two files with content e.g. body 'x' works OK

A workaround seems to be to comment the [100-continue line](https://github.com/boto/boto/issues/840#issuecomment-812257064) out of boto3 - but I'm not sure if that is healthy when switching back to the Amazon s3 endpoint again.

Whilst I get that the minio server and the native python `minio` package might be more 'correct', one could argue that minio is not simulating s3 properly, if the Amazon s3 endpoint works ok with boto3, but minio does not. 

I also get that its incredibly frustrating that boto3 won't fix their stuff.

**Comment by harshavardhana** [comment]
> @harshavardhana Interesting, thanks. This of course makes using `boto3` + `minio` s3 unreliable. I just ran into an even simpler repro case of the 60 sec freeze - creating an empty file followed by file with some content:

There is nothing we can do @abulka you can use our minio-py SDK which does not suffer from this problem and it works with AWS S3 without any problems https://github.com/minio/minio-py

**Comment by abulka** [comment]
>  use our minio-py SDK which does not suffer from this problem and it works with AWS S3 without any problems

I hadn't considered this option - not a bad idea.

> There is nothing we can do 

You could add a flag to minio called `--dodgy-s3-100-boto-compatibility-mode` which would allow developers to switch reliably between minio and amazon endpoints using boto3. 😉

**Comment by abulka** [comment]
Just ran into this again today after upgrading boto3 and hitting the same problem (the upgrade overwrote my patch). So I wrote a little module `botocore_patch_detect.py` to detect an unpatched boto3:

```python
# Detect if running unpatched boto3 which has problems running against minio server

from botocore.handlers import add_expect_header


class _Http:
    def get(self, a, b):
        return 'PUT'


class _Model:
    @property
    def http(self):
        return _Http()


class _Body:
    def __init__(self) -> None:
        self.read = ''


def detect_unpatched_boto3():
    model = _Model()
    body = _Body()
    x = model.http.get('method', '')
    headers = {}
    params = {'body': body, 'headers': headers}

    add_expect_header(model, params)
    if 'Expect' in headers and headers['Expect'] == '100-continue':
        raise RuntimeError(
            f"Please hand patch botocore.handlers by commenting out the last line of "
            f"add_expect_header() approx line 297, otherwise you will get 60 sec timeouts "
            f"running against minio")
```

Usage:

```python
from botocore_patch_detect import detect_unpatched_boto3

detect_unpatched_boto3()
```

P.S.

> use our minio-py SDK which does not suffer from this problem and it works with AWS S3 without any problems

An update on this. I'm glad I stayed with boto3 not the minio Library to access minio/amazon because I ended up evolving my code to use the async aware `aioboto3` to make my code asynchronous. As far as I'm aware, there is no async minio python library.
